File: packages/core/examples/playground.ts
import readline from "node:readline/promises";
import { stdin as input, stdout as output } from "node:process";

import { chromium, firefox, setServerAddr, shutdown, type BrowserKind } from "../dist/index.js";

async function main(): Promise<void> {
  const args = parseArgs(process.argv.slice(2));
  setServerAddr(args.serverAddr);

  console.log(
    `[ts-playground] launching ${args.browser} with browserBinary=${JSON.stringify(args.browserBinary)} via singleton TypeScript client runtime`,
  );

  const browserType = args.browser === "firefox" ? firefox : chromium;
  const browser = await browserType.launch({
    browserBinary: args.browserBinary ?? undefined,
  });
  const initialTab = browser.page();
  logBrowserLaunch(browser, initialTab.sessionId);

  const initialNavigation = await initialTab.navigate(args.navigateUrl);
  console.log(
    `[${initialTab.sessionId}] tab navigated: ${initialNavigation.url} (${initialNavigation.note})`,
  );
  logNavigationAutomation(initialTab.sessionId, initialNavigation);

  if (args.clickSelector) {
    const click = await initialTab.click(args.clickSelector);
    console.log(
      `[${initialTab.sessionId}] element clicked: selector=${click.selector} (${click.note}) bidi_session_id=${click.bidiSessionId}`,
    );
  }

  const tabs = [initialTab];
  for (let index = 0; index < args.tabs; index += 1) {
    const tabNumber = index + 2;
    const tab = await browser.newPage();
    console.log(`[${browser.sessionId}] tab opened: ${tab.sessionId} (requested additional tab ${tabNumber})`);
    const navigation = await tab.navigate(args.navigateUrl);
    console.log(`[${tab.sessionId}] tab navigated: ${navigation.url} (${navigation.note})`);
    logNavigationAutomation(tab.sessionId, navigation);
    if (args.clickSelector) {
      const click = await tab.click(args.clickSelector);
      console.log(
        `[${tab.sessionId}] element clicked: selector=${click.selector} (${click.note}) bidi_session_id=${click.bidiSessionId}`,
      );
    }
    tabs.push(tab);
  }

  const rl = readline.createInterface({ input, output });
  await rl.question("[ts-playground] Press Enter to close the tabs and browser session...");
  rl.close();

  for (const tab of tabs.slice(1)) {
    await tab.close();
    console.log(`[${tab.sessionId}] tab session closed`);
  }
  await initialTab.close();
  console.log(`[${initialTab.sessionId}] tab session closed`);

  await browser.close();
  console.log(`[${browser.sessionId}] session closed`);
}

interface ParsedArgs {
  serverAddr: string;
  browser: BrowserKind;
  browserBinary: string | null;
  navigateUrl: string;
  clickSelector: string | null;
  tabs: number;
}

function parseArgs(argv: string[]): ParsedArgs {
  let serverAddr = "127.0.0.1:50051";
  let browser: BrowserKind = "chromium";
  let browserBinary: string | null = null;
  let navigateUrl = "https://example.com";
  let clickSelector: string | null = null;
  let tabs = 3;

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    const next = argv[index + 1];
    if (arg === "--server-addr" && next) {
      serverAddr = next;
      index += 1;
      continue;
    }
    if (arg === "--browser" && next && (next === "chromium" || next === "firefox")) {
      browser = next;
      index += 1;
      continue;
    }
    if (arg === "--browser-binary" && next) {
      browserBinary = next;
      index += 1;
      continue;
    }
    if (arg === "--navigate-url" && next) {
      navigateUrl = next;
      index += 1;
      continue;
    }
    if (arg === "--click-selector" && next) {
      clickSelector = next;
      index += 1;
      continue;
    }
    if (arg === "--tabs" && next) {
      tabs = Number.parseInt(next, 10);
      index += 1;
      continue;
    }
  }

  return {
    serverAddr,
    browser,
    browserBinary,
    navigateUrl,
    clickSelector,
    tabs,
  };
}

function logBrowserLaunch(browser: { sessionId: string; browserName: string; launchNote: string; cdpWebSocketURL: string; userDataDir: string }, initialTabSessionId: string): void {
  const cdpPart = browser.cdpWebSocketURL ? ` cdp=${browser.cdpWebSocketURL}` : "";
  console.log(
    `[${browser.sessionId}] browser launched: ${browser.browserName} (${browser.launchNote})${cdpPart} user_data_dir=${browser.userDataDir} initial_tab_session_id=${initialTabSessionId}`,
  );
}

function logNavigationAutomation(
  tabSessionId: string,
  navigation: { bidiSessionId: string; mapperTargetId: string; mapperSessionId: string; packageVersion: string },
): void {
  console.log(
    `[${tabSessionId}] automation session: bidi_session_id=${navigation.bidiSessionId} mapper_target_id=${navigation.mapperTargetId} mapper_session_id=${navigation.mapperSessionId} package_version=${navigation.packageVersion}`,
  );
}

main()
  .catch(async (error: unknown) => {
    console.error(error);
    process.exitCode = 1;
  })
  .finally(async () => {
    await shutdown();
  });

File: rust/allwright/examples/playground.rs
use allwright::{BrowserKind, LaunchOptions, launch_browser, set_server_addr, shutdown};
use clap::Parser;
use std::io;

#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
struct Args {
    /// Engine server address, with or without scheme
    #[arg(long, default_value = "127.0.0.1:50051")]
    server_addr: String,

    /// Browser backend to launch: chromium or firefox
    #[arg(long, default_value = "chromium")]
    browser: String,

    /// Optional browser binary path or executable name override
    #[arg(long)]
    browser_binary: Option<String>,

    /// Number of additional tabs to open after the initial browser tab
    #[arg(long, default_value_t = 3)]
    tabs: u8,

    /// URL to navigate each opened tab to
    #[arg(long, default_value = "https://example.com")]
    navigate_url: String,

    /// Optional CSS selector to click over BiDi after navigation
    #[arg(long)]
    click_selector: Option<String>,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let args = Args::parse();
    set_server_addr(&args.server_addr)?;
    println!(
        "[playground] launching {} with browser_binary={:?} via singleton Rust client runtime",
        args.browser, args.browser_binary
    );
    let browser_kind = parse_browser_kind(&args.browser)?;
    let browser = launch_browser(
        browser_kind,
        LaunchOptions {
            browser_binary: args.browser_binary.clone(),
            timeout_ms: None,
            ..Default::default()
        },
    )
    .await?;

    let initial_tab = browser.initial_tab();
    print_browser_launch(&browser, initial_tab.session_id());

    let initial_navigation = initial_tab.navigate(&args.navigate_url).await?;
    println!(
        "[{}] tab navigated: {} ({})",
        initial_tab.session_id(),
        initial_navigation.url,
        initial_navigation.note
    );
    print_navigation_automation(initial_tab.session_id(), &initial_navigation);

    if let Some(selector) = args.click_selector.as_deref() {
        let click = initial_tab.click(selector).await?;
        println!(
            "[{}] element clicked: selector={} ({}) bidi_session_id={}",
            initial_tab.session_id(),
            click.selector,
            click.note,
            click.bidi_session_id
        );
    }

    let mut additional_tabs = Vec::new();
    for tab_index in 0..args.tabs {
        let tab_number = usize::from(tab_index) + 2;
        let tab = browser.new_tab().await?;
        println!(
            "[{}] tab opened: {} (requested additional tab {})",
            browser.session_id(),
            tab.session_id(),
            tab_number
        );

        let navigation = tab.navigate(&args.navigate_url).await?;
        println!(
            "[{}] tab navigated: {} ({})",
            tab.session_id(),
            navigation.url,
            navigation.note
        );
        print_navigation_automation(tab.session_id(), &navigation);

        if let Some(selector) = args.click_selector.as_deref() {
            let click = tab.click(selector).await?;
            println!(
                "[{}] element clicked: selector={} ({}) bidi_session_id={}",
                tab.session_id(),
                click.selector,
                click.note,
                click.bidi_session_id
            );
        }

        additional_tabs.push(tab);
    }

    wait_for_enter(
        "[playground] Press Enter to close the tabs and browser session so the browser stays open for observation...",
    )
    .await?;

    for tab in &additional_tabs {
        tab.close().await?;
        println!("[{}] tab session closed", tab.session_id());
    }
    initial_tab.close().await?;
    println!("[{}] tab session closed", initial_tab.session_id());

    browser.close().await?;
    println!("[{}] session closed", browser.session_id());

    shutdown().await;
    Ok(())
}

async fn wait_for_enter(prompt: &str) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    println!("{prompt}");
    tokio::task::spawn_blocking(|| {
        let mut line = String::new();
        io::stdin().read_line(&mut line)?;
        Ok::<(), io::Error>(())
    })
    .await??;
    Ok(())
}

fn parse_browser_kind(
    value: &str,
) -> Result<BrowserKind, Box<dyn std::error::Error + Send + Sync>> {
    match value.trim().to_ascii_lowercase().as_str() {
        "chromium" | "chrome" => Ok(BrowserKind::Chromium),
        "firefox" => Ok(BrowserKind::Firefox),
        other => {
            Err(format!("unsupported --browser value `{other}`; use chromium or firefox").into())
        }
    }
}

fn print_browser_launch(browser: &allwright::Browser, initial_tab_session_id: &str) {
    let cdp = browser.cdp_websocket_url();
    if cdp.is_empty() {
        println!(
            "[{}] browser launched: {} ({}) user_data_dir={} initial_tab_session_id={}",
            browser.session_id(),
            browser.browser_name(),
            browser.launch_note(),
            browser.user_data_dir(),
            initial_tab_session_id,
        );
    } else {
        println!(
            "[{}] browser launched: {} ({}) cdp={} user_data_dir={} initial_tab_session_id={}",
            browser.session_id(),
            browser.browser_name(),
            browser.launch_note(),
            cdp,
            browser.user_data_dir(),
            initial_tab_session_id,
        );
    }
}

fn print_navigation_automation(tab_session_id: &str, navigation: &allwright::NavigateResult) {
    println!(
        "[{}] automation session: bidi_session_id={} mapper_target_id={} mapper_session_id={} package_version={}",
        tab_session_id,
        navigation.bidi_session_id,
        navigation.mapper_target_id,
        navigation.mapper_session_id,
        navigation.package_version
    );
}
